05-part_two.Rmd

```diff
@@ -30,15 +34,72 @@ get_max(c("a", "b", "c"))
+## Formalised testing
+Beyond in-script tests, an analyst may decide to take a more formal approach to testing if the project warrants it. Software development has dedicated testing professionals (called "testers") and there is plenty of literature about software testing on the internet. Much of this is transferable to code based analysis, but not all of it. There are times when it makes sense to make a distinction between a model (which is generally operated by a technical specialist) and an IT product (e.g. a website, which is used by a broader user base). Shiny apps sit somewhere between a model and an IT product so we'll include some specific comments about these.
+
+There are often four types of software testing covered:
+
+* Unit testing (tests the different components or "units" of code using tools such as `testthat` in R).
+* Integration testing (tests the correct flow of data through the components).
+* System testing (tests the overall outputs make sense).
+* Acceptance testing (tests whether or not the work meets its objectives).
+
+Software testing can make a distinction between functional and non-functional testing too. In the context of an analytical model, a functional test would be one which tests the model produces what its intended to whereas a non-functional test would cover things such as its run-time and interface (i.e. things which affect the experience of using it, but not the numbers that come out).
+
+For more information, see [here](https://www.guru99.com/levels-of-testing.html) or use your favourite search engine to explore the topic.
+
+As this is a chapter about testing code, we are only going to cover unit testing in more detail. The other testing aspects cover testing of wider areas of a software project than just the code, and other parts of the DfE QA framework cover how to handle these:
+
+* Integration testing tests correct data flow, transformations and interactions between different model components.
+* System testing relates to verification tests on the overall outputs.
+* Acceptance testing relates to the validation pillar of the QA framework.
+
+### Unit testing
+
+Unit testing seeks to test every distinct component within your code to enable to you isolate specific areas that may be introducing errors. If you have written you code using functions, it is sometimes easier to think of this as testing each specific function.
+
+One of the key principles of unit testing is that each unit (or function) should be tested independently of all the others and without dependency on any data sources that may change. The objective is to isolate the cause of any errors:
+
+* If you test a function relying on output from previous functions and that test throws an error, you do not immediately know which of the functions (or units) caused the error and therefore it is not a unit test.
```

Comment:
As line 60, this needs rephrasing to remove the you/your references, eg: "If a test relies on an output from a previous function and it throes an error, it is not obvious which of the functions caused the error